### PR TITLE
 Implemented aggregate values instead of the keys in the Evolution dashboard URL

### PR DESCRIPTION
**Pull request by ab0092**

### CODE REVIEW

**Comment by chutten** [comment]
Ah-ha, I knew I was missing something.

Sorry for the delay, I was out of the office last week. I'll take a look at this today.

**Comment by ab0092** [comment]
No issues. Please have a look and let me know if any tweaks are required.. Thanks.

File: new-pipeline/src/evo.js [hunk 1]
```diff
@@ -412,8 +412,14 @@ function updateAggregates(kind, buckets) {
   }
 
   // Load aggregates from state on first load
-  newAggregates = newAggregates.map(entry => entry[0]);
-  let aggregates = gInitialPageState.aggregates.filter(aggregate => newAggregates.includes(aggregate));
+  let aggregates = gInitialPageState.aggregates
```
Comment (chutten) [inline new-pipeline/src/evo.js]: Could you provide an explanation for what transformation you're performing here? It seems as though you're taking the initial aggregates and... I'm actually a little lost already.

File: new-pipeline/src/evo.js [hunk 2]
```diff
@@ -412,8 +412,14 @@ function updateAggregates(kind, buckets) {
...
+                    .filter((aggregate)=>{
+                      return newAggregates.filter((newaggregate)=> newaggregate[1] == aggregate).length > 0
```
Comment (chutten) [inline new-pipeline/src/evo.js]: This code filters based on aggregates who are present as the second index in one of the newAggregates? Perhaps this will be more clear?

`.filter(aggregate => newAggregates.some(newAggregate => newAggregate[1] === aggregate))`

**Review by chutten — COMMENTED** [review]
Is it possible to write this code in a backward-compatible manner to ensure that old links people have lying around will still work?

The style in the file is a little eclectic, but mostly we have spaces before and after `=>` and after `,`

Looks like a good first draft!

File: new-pipeline/src/evo.js (hunk 2 above)
Comment (ab0092) [reply new-pipeline/src/evo.js]: The transformation happening here is actually from the values to keys because as we are now storing the values of aggregates in the url instead of the keys so while restoring data back into to UI i have remapped the values to the keys for the multi-select control to work as required.

Comment (ab0092) [reply new-pipeline/src/evo.js]: And you are right at the 0th index the keys & at the 1st index the values are stored respectively.

**Comment by ab0092** [comment]
@chutten i have implemented the backward compatibility option and also formatted the code accordingly. Please have a look if anything more is to be added.

File: new-pipeline/src/dashboards.js [hunk 3]
```diff
@@ -227,8 +227,8 @@ function loadStateFromUrlAndCookie() {
     var aggregates = pageState.aggregates.split("!")
       .filter(function (v) {
         return ["5th-percentile", "25th-percentile", "median",
-            "75th-percentile", "95th-percentile", "mean", "submissions"].indexOf(v) >= 0 ||
-          v.startsWith("bucket-");
+            "75th-percentile", "95th-percentile", "mean", "submissions"].indexOf(v) >= 0
+            ||(v.trim() != "");
```
Comment (chutten) [inline new-pipeline/src/dashboards.js]: Need a space after `||`, or you can leave it at the end of the previous line.

File: new-pipeline/src/evo.js (hunk 1 above)
Comment (chutten) [reply new-pipeline/src/evo.js]: I think I get it now. We want a list of newAggregates' keys where either the key or the value of the newAggregate is present in gInitialPageState.aggregates. Is that right? I think we can make this a bit more obvious if we write it something like:
```js
let aggregates = [];
for (let [key, value] of newAggregates) {
  if (gInitialPageState.aggregates.includes(key) || gInitialPageState.aggregates.includes(value)) {
    aggregates.push(key);
  }
}
```

**Review by chutten — COMMENTED** [review]
Do I have the logical transformation correct here? It's tricky to unwrap the functional transformations needed, so I'm not sure I have it right.

File: new-pipeline/src/dashboards.js (hunk 3 above)
Comment (ab0092) [reply new-pipeline/src/dashboards.js]: The changes made here is actually done to make the old url's to also work so i have added the change of checking for only non-empty values and later on check the validity of the values in the savestateToCookieandUrl methods

**Comment by ab0092** [comment]
@chutten i have added the required changes. And thanks for the above code snippet idea. It really makes it more understandable and clean.  :-)

File: new-pipeline/src/evo.js [hunk 4]
```diff
@@ -921,10 +926,14 @@ function saveStateToUrlAndCookie() {
     cumulative = gInitialPageState.cumulative,
     trim = gInitialPageState.trim,
     sortKeys = gInitialPageState.sort_keys,
-    selectedKeys = gInitialPageState.keys;
+    selectedKeys = gInitialPageState.keys,
+    aggregateOptions = $("#aggregates option"),
+    // Get Aggregate Values instead of keys to save 
+    // in the state url to increase readability
+    aggregateValues = $("#aggregates").val()
```
Comment (chutten) [inline new-pipeline/src/evo.js:933]: From the previous code it seems possible that `$("#aggregates").val()` could sometimes fail to return the selected aggregates (thus the need for the `|| []`). Don't we need to check for that case here, too?

**Review by chutten — COMMENTED** [review]
Looks mostly good, though I think I may have noticed a corner case.

How did you test this? Is there a live preview somewhere I can play with?

File: new-pipeline/src/evo.js (hunk 4 above)
Comment (ab0092) [reply new-pipeline/src/evo.js:933]: You can check the application here https://ab0092.github.io/telemetry-dashboard/

And for the validation as per my understanding we do not require the extra check as all the items in the multi-select control cannot be unchecked and the first option is always kept checked so may be we do not require it here.

Comment (chutten) [reply new-pipeline/src/evo.js:933]: Interestingly, you are correct. When you remove the final option's checkmark it resets to the initial or default selection (which can be the first aggregate, or some combination of a few of them) which keeps us clear of errors here.

(( this reverting to default selection is probably a bug, but we can file and fix that later. ))

**Comment by chutten** [comment]
Looks good, works well, and I confirm that old-type "bucket-0!bucket-4"-type urls [like this one](https://ab0092.github.io/telemetry-dashboard/new-pipeline/evo.html#!aggregates=bucket-0!bucket-4&cumulative=0&end_date=2018-08-19&keys=!__none__!__none__&max_channel_version=nightly%252F63&measure=TELEMETRY_SEND_FAILURE_TYPE&min_channel_version=nightly%252F60&processType=*&product=Firefox&sanitize=1&sort_keys=submissions&start_date=2018-06-25&trim=1&use_submission_date=0) are compatible with the new logic.

Good work!

**Review by chutten — APPROVED** [review]

**Comment by chutten** [comment]
And that should be that. Have an eye on what you'd like to work on next?

**Comment by ab0092** [comment]
@chutten i have not yet been able to look at the open issues in details so could you suggest one which is having an higher priority ?? Or else i will have a look at the open tickets and decide.

**Comment by chutten** [comment]
Take your pick, there are none that are immediately pressing at this moment.

**Comment by ab0092** [comment]
#570 is an issue i had inquired about but did not receive any update so if you could provide me some details then we are good to go. We can discuss the rest on that thread itself.
